### PR TITLE
[CODE] Fill out completion spec for code (VS Code)

### LINKED ISSUE #210: [CODE] Complete VS Code spec
### CLI Tool

The `code` spec is very minimal right now

### PR DESCRIPTION
**Pull request by Tyriar**
This implements the whole spec for `code` based on `code --help`

Fixes #210

### CODE REVIEW

File: dev/code.ts [hunk 1]
```diff
@@ -1,40 +1,221 @@
...
+  {
+    name: ["-d", "--diff"],
+    description: "Compare two files with each other.",
+    args: [
+      {
+        template: "filepaths",
+      },
+      {
+        template: "filepaths",
+      },
```
Comment (QuiiBz) [inline dev/code.ts:11]: I think we should add `name` and `description` to all arguments, to get better feedback in the autocomplete box. Same for all other `args` objects.

**Review by QuiiBz — COMMENTED** [review]
Thanks for your contribution!

File: dev/code.ts [hunk 2]
```diff
@@ -1,40 +1,228 @@
...
         {
-          template: "filepaths",
+          name: "--category",
+          description: "Filters installed extensions by provided category",
         },
         {
-          template: "filepaths",
+          name: "--show-versions",
+          description:
+            "Show versions of installed extensions, when using --list-extensions.",
```
Comment (cstrnt) [inline dev/code.ts:106]: Hey there, this might sound weird, but those are `options`and not `args`. We don't offer the option to have conditional `options`

Comment (cstrnt) [reply dev/code.ts:106]: As a rule of thumb: If it starts with `-` or `--` it's an option

File: dev/code.ts [hunk 3]
```diff
@@ -1,40 +1,228 @@
...
+    description: "Print version.",
+  },
+  {
+    name: "--verbose",
+    description: "Print verbose output (implies --wait).",
+  },
+  {
+    name: "--log",
+    description: "Log level to use. Default is 'info'.",
+    args: {
```
Comment (cstrnt) [inline dev/code.ts:147]: This should be option because it has a default value if omited

File: dev/code.ts [hunk 4]
```diff
@@ -1,40 +1,228 @@
...
+      name: "port",
+    },
+  },
+  {
+    name: "--disable-gpu",
+    description: "Disable GPU hardware acceleration.",
+  },
+  {
+    name: "--max-memory",
+    description: "Max memory size for a window (in Mbytes).",
```
Comment (cstrnt) [inline dev/code.ts:209]: This should take an argument

File: dev/code.ts [hunk 5]
```diff
@@ -1,40 +1,228 @@
...
+  {
+    name: "--list-extensions",
+    description: "List the installed extensions.",
+    args: {
+      isOptional: true,
+      suggestions: [
         {
-          template: "filepaths",
+          name: "--category",
+          description: "Filters installed extensions by provided category",
```
Comment (cstrnt) [inline dev/code.ts:101]: this should take an argument, right?

**Review by cstrnt — CHANGES_REQUESTED** [review]
Hey there 👋 
thank you very much for you contribution :) I just left some comments, but apart from that that's a really nice PR for a really great IDE 😉

File: dev/code.ts (hunk 3 above)
Comment (Tyriar) [reply dev/code.ts:147]: The wording is a little awkward, but I think this is correct as mandatory. It's calling out the default log level if you do not specify `--log`

File: dev/code.ts (hunk 2 above)
Comment (Tyriar) [reply dev/code.ts:106]: How do I specify nested options? These 2 are only valid when --list-extensions is provided

File: dev/code.ts (hunk 5 above)
Comment (Tyriar) [reply dev/code.ts:101]: Yep, pending nesting options question

File: dev/code.ts (hunk 2 above)
Comment (Tyriar) [reply dev/code.ts:106]: If these don't exist I can just make it top level

**Comment by Tyriar** [comment]
Fixed this up, also added suggestions for `--category`

File: dev/code.ts (hunk 2 above)
Comment (cstrnt) [reply dev/code.ts]: There are no nested options as of right now :/

File: dev/code.ts [hunk 6]
```diff
@@ -1,40 +1,253 @@
...
+      },
+    ],
+  },
+  {
+    name: ["-a", "--add"],
+    description: "Add folder(s) to the last active window",
+    args: {
+      name: "folder",
+      template: "folders",
+    },
```
Comment (QuiiBz) [inline dev/code.ts:22]: Shouldn't this be `variadic`? Since the description contains "folder(s)".

Comment (Tyriar) [reply dev/code.ts:22]: I added `variadic: true` but the file template is still showing up?

Screen Shot 2021-05-22 at 4 50 31 AM
Screen Shot 2021-05-22 at 4 50 37 AM

Comment (QuiiBz) [reply dev/code.ts:22]: [Variadic](https://fig.io/docs/autocomplete/schemas/arg#variadic) means that an argument repeats infinitely, so it should autocomplete folders infinitely. It does not work?

Comment (Tyriar) [reply dev/code.ts]: Yeah it didn't work, it may be a bug with fig? I committed the change regardless as it seems like it's correct.

Comment (cstrnt) [reply dev/code.ts]: That's really weird and seems to be a bug! We'll investigate!

**Comment by cstrnt** [comment]
@Tyriar 
Hey there, if you like you could merge in the current master and remove the `code.js` from the git index (`git rm -f --cached specs/code.js`:) I think this PR is ready to get merged though, because it's a Bug with Fig itself and not your Spec!

**Review by QuiiBz — APPROVED** [review]
Thanks for this PR, ready to merge @cstrnt

**Comment by cstrnt** [comment]
Opened #219

**Pull request merged**